usr/src/cmd/pkgsrc-wrapper/pkgsrc-wrapper.c

```diff
@@ -0,0 +1,181 @@
+static struct {
+	const char *bin_name;
+	const char *bin_needs;
+} remap_table[] = {
+	{ "csh", "tcsh" },
+	/* login(1) invokes shells as -$SHELL. Cover that here. */
+	{ "-csh", "tcsh" },
+	{ "-tcsh", "tcsh" },
+	{ "-zsh", "zsh" },
+	{ NULL, NULL}
+};
+
+static const char *
+remap_bin(const char *base_bin)
+{
+	uint_t index;
+
+	for (index = 0; remap_table[index].bin_name != NULL; index++) {
+		/* Can use strcmp() because remap_table entries are bounded. */
+		if (strcmp(remap_table[index].bin_name, base_bin) == 0)
+			return (remap_table[index].bin_needs);
+
+	}
+	return (base_bin);
+}
+
+
+/* List these pkgsrc paths in desired search order. */
+static const char *pkgsrc_paths[] = {
+	"/opt/local/bin",
+	"/opt/local/sbin",
+	"/opt/tools/bin",
+	"/opt/tools/sbin",
+	NULL
+};
+
+/*
+ * This returns an allocated string.  Since we're exec()-ing or exit()-ing we
+ * really don't need to worry about leaks, however.
+ */
+static char *
+generate_pkgsrc_path(const char *desired_bin)
+{
+	const char *path = NULL;
+	char *binpath;
+	uint_t index;
+
+	for (index = 0; pkgsrc_paths[index] != NULL; index++) {
+		int dirfd;
+		int cmdfd;
+
+		/*
+		 * NOTE: Any of dirfd or cmdfd that has an actual file
+		 * descriptor from open(2) MUST be closed before this loop
+		 * either exits or iterates.
+		 */
+
+		dirfd = open(pkgsrc_paths[index], O_RDONLY);
+		if (dirfd == -1) {
+			/* XXX KEBE ASKS, print diagnostics? */
+			continue; /* Try next one */
+		}
+		cmdfd = openat(dirfd, desired_bin, O_RDONLY);
+		(void) close(dirfd); /* XXX KEBE WARNS, errno reset... */
+		if (cmdfd == -1) {
+			/*
+			 * XXX KEBE ASKS, print diagnostics?
+			 * If so, move close(dirfd) here after printing, or
+			 * save off errno for openat() above.
+			 */
+			continue; /* Try next one */
+		}
+		(void) close(cmdfd);
+		path = pkgsrc_paths[index];
+		break;	/* We're done. */
+	}
+
+	if (path == NULL)
+		return (NULL); /* If we reach here, there's no such binary. */
+
+	/*
+	 * Okay, so we can now combine path and desired_bin into a string
+	 * suitable for an exec()-family call.
+	 */
+
+	if (asprintf(&binpath, "%s/%s", path, desired_bin) == -1) {
+		/* XXX KEBE ASKS, print diagnostics? */
+		assert(binpath == NULL); /* man pages says this is true. */
+	}
+
+	return (binpath);
+}
+
+/* Don't let us run ourselves. */
+static char *myname = "pkgsrc-wrapper";
+
+int
+main(int argc, char *argv[], char *envp[])
+{
+	const char *base_bin;
+	const char *remapped_bin;
+	char *pkgsrc_path;
+
+	/*
+	 * 1. Get the binary name in argv[0], remapping if need be.
+	 * (e.g. csh -> tcsh, or -($SHELL) invocations.)
+	 */
+	base_bin = strrchr(argv[0], '/');
+	if (base_bin == NULL)
+		base_bin = argv[0];
+	else
+		base_bin++; /* Move past matched '/' */
+
+	/*
+	 * 2. Make sure we just aren't running ourselves!
+	 * (Can use strcmp() because "myname" is bounded.)
+	 */
+	if (strcmp(myname, base_bin) == 0) {
+		/* XXX KEBE ASKS L10N? */
+		(void) fprintf(stderr, "%s should not be run by itself.\n",
```

Comment:
Change this to "%s should not be run directly.\n"

As far as L10N, I'd like to sure, but how many languages do we need before it's satisfactory? And how long will it take to get proper translations? But on the other hand, there's no l10n in things like vmadm...

Reply:
We'll leave it out then, for now.

Reply:
Also, will fix.